File: csharp/exercise/exer06b02-blocking-queue.cs
/*
 * BLOCKING QUEUE IMPLEMENTATION
 * Version B02: General blocking queues
 *              Underlying mechanism: Condition variables
 */
using System;
using System.Collections.Generic;
using System.Threading;



class Exer06B02 : IRunnable
{
    public void run()
    {
        var queue = new MyBlockingQueue<string>(2); // capacity = 2
        new Thread(() => producer(queue)).Start();
        new Thread(() => consumer(queue)).Start();
    }


    private void producer(MyBlockingQueue<string> queue)
    {
        string[] arr = { "nice", "to", "meet", "you" };

        foreach (var data in arr)
        {
            queue.put(data);
            Console.WriteLine($"Producer: {data}\t\t\t[done]");
        }
    }


    private void consumer(MyBlockingQueue<string> queue)
    {
        string data;
        Thread.Sleep(5000);

        for (int i = 0; i < 4; ++i)
        {
            data = queue.take();
            Console.WriteLine($"\tConsumer: {data}");

            if (0 == i)
                Thread.Sleep(5000);
        }
    }


    //////////////////////////////////////////////


    class MyBlockingQueue<T>
    {
        private object condEmpty = new object();
        private object condFull = new object();

        private int capacity = 0;
        private Queue<T> queue = null;


        public MyBlockingQueue(int capacity)
        {
            if (capacity <= 0)
                throw new ArgumentException("capacity must be a positive integer");

            this.capacity = capacity;
            queue = new Queue<T>();
        }


        public void put(T value)
        {
            lock (condFull)
            {
                while (queue.Count >= capacity)
                {
                    // Queue is full, must wait for 'take'
                    Monitor.Wait(condFull);
                }

                lock (queue)
                {
                    queue.Enqueue(value);
                }
            }

            lock (condEmpty)
            {
                Monitor.Pulse(condEmpty);
            }
        }


        public T take()
        {
            T result = default;

            lock (condEmpty)
            {
                while (0 == queue.Count)
                {
                    // Queue is empty, must wait for 'put'
                    Monitor.Wait(condEmpty);
                }

                lock (queue)
                {
                    result = queue.Dequeue();
                }
            }

            lock (condFull)
            {
                Monitor.Pulse(condFull);
            }

            return result;
        }
    }
}

File: python/exer06b02_blocking_queue.py
'''
BLOCKING QUEUE IMPLEMENTATION
Version B02: General blocking queues
             Underlying mechanism: Condition variables
'''

import time
import threading



class BlockingQueue:
    def __init__(self, capacity: int):
        if capacity <= 0:
            raise ValueError('capacity must be a positive integer')
        self.__capacity = capacity
        self.__lk = threading.Lock()
        self.__cond_empty = threading.Condition(self.__lk)
        self.__cond_full = threading.Condition(self.__lk)
        self.__q = [] # queue


    def put(self, value):
        with self.__lk:
            while len(self.__q) >= self.__capacity:
                self.__cond_full.wait()

            self.__q.append(value)
            self.__cond_empty.notify()


    def take(self):
        result = None

        with self.__lk:
            while len(self.__q) == 0:
                self.__cond_empty.wait()

            result = self.__q.pop(0)
            self.__cond_full.notify()

        return result



def producer(q: BlockingQueue):
    arr = [ 'nice', 'to', 'meet', 'you' ]

    for value in arr:
        print(f'Producer: {value}')
        q.put(value)
        print(f'Producer: {value} \t\t\t[done]')



def consumer(q: BlockingQueue):
    time.sleep(5)

    for i in range(4):
        value = q.take()
        print(f'\tConsumer: {value}')

        if i == 0:
            time.sleep(5)



blkqueue = BlockingQueue(2) # capacity = 2
threading.Thread(target=producer, args=(blkqueue,)).start()
threading.Thread(target=consumer, args=(blkqueue,)).start()
